Code/Framework/AzToolsFramework/AzToolsFramework/Thumbnails/ThumbnailerComponent.cpp

```diff
@@ -49,59 +53,114 @@ namespace AzToolsFramework
+        void ThumbnailerComponent::RegisterThumbnailProvider(SharedThumbnailProvider provider)
         {
-            AZ_Assert(m_thumbnails.find(contextName) == m_thumbnails.end(), "Context %s already registered", contextName);
-            m_thumbnails[contextName] = AZStd::make_shared<ThumbnailContext>();
-        }
+            auto it = AZStd::find_if(m_providers.begin(), m_providers.end(), [provider](const SharedThumbnailProvider& existingProvider)
+                {
+                    return AZ::StringFunc::Equal(provider->GetProviderName(), existingProvider->GetProviderName());
+                });
 
-        void ThumbnailerComponent::UnregisterContext(const char* contextName)
-        {
-            AZ_Assert(m_thumbnails.find(contextName) != m_thumbnails.end(), "Context %s not registered", contextName);
-            m_thumbnails.erase(contextName);
+            if (it != m_providers.end())
+            {
+                AZ_Error("ThumbnailerComponent", false, "Provider with name %s is already registered with context.", provider->GetProviderName());
+                return;
+            }
+
+            m_providers.insert(provider);
         }
 
-        bool ThumbnailerComponent::HasContext(const char* contextName) const
+        void ThumbnailerComponent::UnregisterThumbnailProvider(const char* providerName)
         {
-            return m_thumbnails.find(contextName) != m_thumbnails.end();
+            AZStd::erase_if(
+                m_providers,
+                [providerName](const SharedThumbnailProvider& provider)
+                {
+                    return AZ::StringFunc::Equal(provider->GetProviderName(), providerName);
+                });
+
         }
 
-        void ThumbnailerComponent::RegisterThumbnailProvider(SharedThumbnailProvider provider, const char* contextName)
+        SharedThumbnail ThumbnailerComponent::GetThumbnail(SharedThumbnailKey key)
         {
-            auto it = m_thumbnails.find(contextName);
-            AZ_Assert(it != m_thumbnails.end(), "Context %s not registered", contextName);
-            it->second->RegisterThumbnailProvider(provider);
+            // find provider who can handle supplied key
+            for (auto& provider : m_providers)
+            {
+                SharedThumbnail thumbnail;
+                if (provider->GetThumbnail(key, thumbnail))
+                {
+                    // if thumbnail is ready return it
+                    if (thumbnail->GetState() == Thumbnail::State::Ready)
+                    {
+                        return thumbnail;
+                    }
+
+                    // if thumbnail is not loaded, start loading it, meanwhile return loading thumbnail
+                    if (thumbnail->GetState() == Thumbnail::State::Unloaded)
+                    {
+                        // listen to the loading signal, so the anyone using it will update loading animation
+                        AzQtComponents::StyledBusyLabel* busyLabel;
```

Comment:
I'm pretty sure this is existing logic that's been ported over, but just so to make sure I'm reading this right: we have a singleton busy label widget that we use to render all of our still-loading thumbnails?

Reply:
I believe that is the case. If not, it should be. We only need one instance of the animation that can be painted anywhere there is a pending thumbnail.

All of the logic was moved almost verbatim from the thumbnail context class into the thumbnail component class. 

There is another pre-existing problem, likely within this block of code, that needs to be debugged. The loading animation persists even after a thumbnail has finished loading. it updates to the correct, loaded thumbnail image after a mouse over or some other interaction.